src/dataframe.rs

```diff
@@ -451,6 +458,40 @@ impl PyDataFrame {
         Ok(table)
     }
 
+    fn __arrow_c_stream__<'py>(
+        &'py mut self,
+        py: Python<'py>,
+        requested_schema: Option<Bound<'py, PyCapsule>>,
+    ) -> PyResult<Bound<'py, PyCapsule>> {
+        let mut batches = wait_for_future(py, self.df.as_ref().clone().collect())?;
+        let mut schema: Schema = self.df.schema().to_owned().into();
+
+        if let Some(schema_capsule) = requested_schema {
+            validate_pycapsule(&schema_capsule, "arrow_schema")?;
+
+            let schema_ptr = unsafe { schema_capsule.reference::<FFI_ArrowSchema>() };
+            let desired_schema = Schema::try_from(schema_ptr).map_err(DataFusionError::from)?;
```

Comment:
Ah right, `Schema::from_pyarrow_bound` accepts an _object_ with `__arrow_c_schema__`, not the raw pointer itself. This is why in pyo3-arrow I expose helper functions to work with the raw capsules.